server/src/main/java/org/elasticsearch/action/admin/cluster/allocation/DesiredBalanceResponse.java

```diff
@@ -171,45 +171,35 @@ public record ShardView(
         String node,
         boolean nodeIsDesired,
         @Nullable String relocatingNode,
-        boolean relocatingNodeIsDesired,
+        @Nullable Boolean relocatingNodeIsDesired,
```

Comment:
I wonder if we really need this to be a `@Nullable Boolean` with its trappy NPE risks. It's `null` if and only if `relocatingNode` is `null`, and the only place where it matters is the XContent output, so it seems like it would be nicer to keep this as a bare `boolean`.

In any case, it'd be good to have an assertion to document/verify the relationship between `relocatingNode` and `relocatingNodeIsDesired`: as written, they should both be null or non-null, and as proposed `relocatingNodeIsDesired` should be `false` if `relocatingNode` is null.

Reply:
I think it is possible to only update `toXContent` to emit null `relocating_node_is_desired` only if `relocating_node`,
however I would prefer to keep the state of the object aligned to its toXContent/toString. This should also simplify testing a little.

+1 for adding the assertion.

Reply:
Actually, assertion might fail if we are reading from the older version